src/types/object-node.ts

```diff
@@ -188,15 +188,27 @@ export class ObjectNode extends Node {
 export function createObjectFactory<S extends Object, T extends S>(baseModel: T): IModelFactory<S, T>
 export function createObjectFactory<S extends Object, T extends S>(name: string, baseModel: T): IModelFactory<S, T>
 export function createObjectFactory(arg1, arg2?) {
-    return createFactory(
-        typeof arg1 === "string" ? arg1 : "unnamed-object-factory",
-        ObjectNode,
-        {
-            isObjectFactory: true,
-            baseModel: typeof arg1 === "string" ? arg2 : arg1
-        },
-        () => observable.shallowObject({})
+    let name = typeof arg1 === "string" ? arg1 : "unnamed-object-factory"
+    let config: Object = typeof arg1 === "string" ? arg2 : arg1
+    let primitiveKeys = Object.keys(config).filter(key => isPrimitive(config[key]))
+
+    let factory = createFactory(
+        name,
+        "object",
+        snapshot => typeof snapshot === 'object' && Object.keys(snapshot).length === primitiveKeys.length && primitiveKeys.every(key => key in snapshot),
```

Comment:
I think the subtypes of each key/ value should also be checked to be fully correct? E.g. there could be a `unionOf({ x: 3 }, { x: "hello")`